### PR TITLE
refactor!: delete insertmode

### PR DESCRIPTION
**Pull request by gpanders**
Neovim already removed `evim` (or any similar flags). The `'insertmode'`
option is a weird remnant, so get rid of it.

### CODE REVIEW

File: src/nvim/options.lua [hunk 1]
```diff
@@ -1248,13 +1248,6 @@ return {
       varname='p_inf',
       defaults={if_true=false}
     },
-    {
-      full_name='insertmode', abbreviation='im',
-      short_desc=N_("start the edit of a file in Insert mode"),
-      type='bool', scope={'global'},
-      varname='p_im',
-      defaults={if_true=false}
-    },
```
Comment (bfredl) [inline src/nvim/options.lua:1257]: This should be kept, but change `varname` to `p_force_off` . compare the entry for `'compatible'`

**Comment by bfredl** [comment]
in principle this is still a breaking change, so commit message should start with `fix!` or `refactor!` .

File: src/nvim/ex_docmd.c [hunk 2]
```diff
@@ -8972,15 +8972,13 @@ bool save_current_state(save_state_T *sst)
   sst->save_restart_edit = restart_edit;
   sst->save_msg_didout = msg_didout;
   sst->save_State = State;
-  sst->save_insertmode = p_im;
```
Comment (zeertzjq) [inline src/nvim/ex_docmd.c:8975]: This item in the struct is unused

**Review by zeertzjq — COMMENTED** [review]
I'm not in favor of this are there people who still use this feature (like #16823), but at least:

**Comment by justinmk** [comment]
> I'm not in favor of this are there people who still use this feature (like #16823), but at least:

@M1cha are you using 'insertmode'? Does this address your use-case instead:

    autocmd VimEnter * startinsert

File: src/nvim/normal.c [hunk 3]
```diff
@@ -1178,14 +1175,6 @@ static void normal_check_stuff_buffer(NormalState *s)
       // if wait_return still needed call it now
       wait_return(false);
     }
-
-    if (need_start_insertmode && goto_im() && !VIsual_active) {
-      need_start_insertmode = false;
```
Comment (justinmk) [inline src/nvim/normal.c:1183]: can `need_start_insertmode` be removed from globals.h ?

**Comment by dundargoc** [comment]
Being accomodating is one thing, but including actual misfeatures is another. I can't think of a single situation where 'insertmode' would be an optimal solution, to anything.

Evim is meant to be opened in a gui so it'd work like a "normal" editor like notepad. In a terminal this just becomes a trap since the escape key won't work, making nvim actually impossible to quit (unless they use ctrl-L, which they can't if they don't already know about it).

That people use this is even worse; it means something in their workflow is fundamentally broken and 'insertmode' is enabling this behavior. I don't usually like telling people how they should use nvim, but in this case I truly believe this can only cause more harm than good.

**Comment by M1cha** [comment]
> > I'm not in favor of this are there people who still use this feature (like #16823), but at least:
> 
> @M1cha are you using 'insertmode'? Does this address your use-case instead:
> 
> ```
> autocmd VimEnter * startinsert
> ```

Yes I am using insertmode because I prefer a "classic" editor but all the others have at least one thing that's annoying and they're not as configurable as vim.
I use CTRL keybindings almost exclusively and don't use normal mode at all(except for search&replace).

Your suggestion would only solve choosing the correct mode when opening vim, but since I use a single, global nvim instance inside nvim-qt rather than opening one instance per file in a terminal I really need insert mode to be the default so whenever I press escape or some command exits by going back to the default mode I need this to be insert-mode rather than normal-mode.

[My current nvim config](https://github.com/M1cha/dotfiles/blob/7294d35/.config/nvim/init.vim)

**Comment by glacambre** [comment]
@M1cha I'm not seeing anything neovim-specific in your config, would switching back to Vim be an option for you?

**Comment by bfredl** [comment]
@M1cha  could this be implemented using mappings?
```
inoremap <esc> <nop>
inoremap <c-l> <esc>
nnoremap <esc> i
```

Most of the internal complexity of `p_im` seems to be behaviours like "when X is used inside a mapping, it behaves like 'insertmode' was not set." However if one makes sure to use noremap mappings all the way, this already works as expected (lhs keys retain their original meaning).

**Comment by M1cha** [comment]
@bfredl not quite. 
- justinmk's autocmd allows starting in insert mode
- your mappings make `esc` go to insert mode and add back `ctrl-l`

but if I e.g. go into normal mode to open a file with `:e filename`, I end up in normal mode instead of insert mode. I guess the issue is that your `<c-l>` switches permanently rather than temporarily.

**Comment by M1cha** [comment]
> @M1cha I'm not seeing anything neovim-specific in your config, would switching back to Vim be an option for you?

Well at the very least it's not a simple drop-in. E.g. vim doesn't support nvim's ipc socket which I use to open new files.

**Comment by clason** [comment]
Well, then you can always stick with Nvim 0.7.x (which will not receive such breaking changes).

**Comment by M1cha** [comment]
>In a terminal this just becomes a trap since the escape key won't work, making nvim actually impossible to quit (unless they use ctrl-L, which they can't if they don't already know about it).

BTW, esc does work just fine in a terminal (for me).

> Well, then you can always stick with Nvim 0.7.x (which will not receive such breaking changes).

nah I don't like using ancient software. If this really ends up being removed I'll have to find another solution. How does evim make vim appear modeless? If they don't use insertmode there's obviously a better way to do this.

**Comment by justinmk** [comment]
@M1cha use-case is valid, we don't want to break it, but let's see if we can get there with a script. Then we can document that.



> but if I e.g. go into normal mode to open a file with `:e filename`, I end up in normal mode instead of insert mode. I guess the issue is that your `<c-l>` switches permanently rather than temporarily.

That can be fixed with a `CmdlineLeave` handler. Building on https://github.com/neovim/neovim/pull/18547#issuecomment-1126679375 , the full script would be:

```
autocmd VimEnter * startinsert
autocmd CmdlineLeave * startinsert
inoremap <esc> <nop>
inoremap <c-l> <esc>
nnoremap <esc> i
```

**Comment by M1cha** [comment]
@justinmk That works pretty well for editing and running commands.
But autocmd (even with just `VimEnter`) seems to break ctrlp. if I press ctrl-p ctrl-f to switch to `mru`, it says `E21: Cannot make changes, 'modifiable' is off`

**Comment by justinmk** [comment]
> That works pretty well ... but seems to break ctrlp.  ctrl-p ctrl-f says `E21`

Do you mean https://github.com/kien/ctrlp.vim ? I guess it's opening a buffer and showing stuff there. Using `silent!` might fix the issue, so now the full script is:

```
autocmd VimEnter,CmdlineLeave,WinEnter,BufEnter * silent! startinsert
inoremap <esc> <nop>
inoremap <c-l> <esc>
nnoremap <esc> i
```

File: src/nvim/testdir/test_edit.vim [hunk 4]
```diff
@@ -1045,8 +1047,8 @@ endfunc
 func Test_edit_F1()
   " Pressing <f1>
   new
-  call feedkeys(":set im\<cr>\<f1>\<c-l>", 'tnix')
```
Comment (zeertzjq) [inline src/nvim/testdir/test_edit.vim:1048]: comment out

File: src/nvim/testdir/test_edit.vim [hunk 5]
```diff
@@ -234,15 +234,17 @@ func Test_edit_09()
   call setline(1, ['abc', 'def', 'ghi'])
   call cursor(1, 1)
   " 1) CTRL-\ CTLR-N
-  call feedkeys(":set im\<cr>\<c-\>\<c-n>ccABC\<c-l>", 'txin')
```
Comment (zeertzjq) [inline src/nvim/testdir/test_edit.vim:237]: comment out

File: src/nvim/testdir/test_edit.vim [hunk 6]
```diff
@@ -16,8 +16,8 @@ func Test_edit_00b()
   call setline(1, ['abc '])
   inoreabbr <buffer> h here some more
   call cursor(1, 4)
-  " <c-l> expands the abbreviation and ends insertmode
-  call feedkeys(":set im\<cr> h\<c-l>:set noim\<cr>", 'tix')
+  " <esc> expands the abbreviation and ends insertmode
```
Comment (zeertzjq) [inline src/nvim/testdir/test_edit.vim:19]:
```suggestion
  " <esc> expands the abbreviation and ends insert mode
```

File: src/nvim/testdir/test_edit.vim [hunk 7]
```diff
@@ -16,8 +16,8 @@ func Test_edit_00b()
   call setline(1, ['abc '])
   inoreabbr <buffer> h here some more
   call cursor(1, 4)
-  " <c-l> expands the abbreviation and ends insertmode
-  call feedkeys(":set im\<cr> h\<c-l>:set noim\<cr>", 'tix')
```
Comment (zeertzjq) [inline src/nvim/testdir/test_edit.vim:20]: comment out

**Comment by M1cha** [comment]
This is the ctrlp I use: https://github.com/ctrlpvim/ctrlp.vim

Your updated script doesn't seem to fix the issue.
A few interesting things:
- In ctrl's list of currently open files, I can see the formatting, e.g. `<bi>#=</bi>`
- when switching between an normal file and a readonly buffer(using my mouse through a split, like `:help`) It does stay in normal mode if I've been in normal mode before switching buffers, but It stays in insert mode without any error if I've been there in the previous buffer

**Comment by bfredl** [comment]
I think we should just do this: 

The remapping and autocmd API:s exposed to plugins are insanely powerful at this point (unlike when they was when ad-hoc options like this were introduced), and it should be possible to emulate 99% of 'insertmode' and if it is not enough, let's give the ecosystem some time and then fill in the remaining 0.99% with API improvements, which will benefit plugins defining custom "experiences" in general, and not just those with hard coded support.

**Comment by justinmk** [comment]
+1. Let's document the current workaround https://github.com/neovim/neovim/pull/18547#issuecomment-1127023354 in the `:help 'insertmode'` replacement, and track the remaining gaps as a backlog item: https://github.com/neovim/neovim/issues/18701

File: runtime/doc/index.txt [hunk 8]
```diff
@@ -1068,7 +1066,7 @@ tag		command		action in Command-line editing mode	~
 |c_<Esc>|	<Esc>		abandon command-line without executing it
 |c_CTRL-[|	CTRL-[		same as <Esc>
 |c_CTRL-\_CTRL-N| CTRL-\ CTRL-N	go to Normal mode, abandon command-line
-|c_CTRL-\_CTRL-G| CTRL-\ CTRL-G	go to mode specified with 'insertmode',
+|c_CTRL-\_CTRL-G| CTRL-\ CTRL-G	go to Normal mode, abandon command-line
 				abandon command-line
```
Comment (bfredl) [inline runtime/doc/index.txt:1070]: duplicate, remove?

**Review by bfredl — APPROVED** [review]
otherwise LGTM, squash?

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
## Resolution

- Removed 'insertmode' option from Nvim. This is an _option_ not a _mode_. This option changes the default behavior of Nvim. We did not remove _insert mode_. 
- We want to drive towards making _emulation_ of the 'insertmode' option possible, with a simple plugin.
- The current "plugin emulation of 'insertmode'" is documented at `:help 'insertmode'`:


```
autocmd VimEnter,CmdlineLeave,WinEnter,WinScrolled,BufEnter * silent! if &modifiable | startinsert | endif
inoremap <Esc> <Nop>
inoremap <C-L> <Esc>
nnoremap <Esc> i
```

---
_Locked to keep the summary visible. You can always [chat](https://neovim.io/#chat) or open a ticket if you have new information/topics to discuss._

**Comment by gpanders** [comment]
@M1cha You may also want to check out https://github.com/tombh/novim-mode

**Pull request merged**